src/countBy.ts

```diff
@@ -0,0 +1,57 @@
+import { purry } from "./purry";
+
+/**
+ * Categorize and count elements in an array using a defined callback function. The transformation function applies to each element in the array to determine its category and then counts how many elements fall into each category.
+ *
+ * @param data - The array.
+ * @param fn - The transformation function.
+ * @signature
+ *   R.countBy(array, fn)
+ * @example
+ *    R.countBy(
+ *      ["a", "b", "c", "B", "A", "a"],
+ *      (x) => x.toLowerCase()
+ *    ) // { a: 3, b: 2, c: 1 }
+ * @dataFirst
+ * @category Array
+ */
+export function countBy<T, K extends PropertyKey>(
+  data: ReadonlyArray<T>,
+  fn: (value: T, index: number, data: ReadonlyArray<T>) => K,
```

Comment:
In `groupBy,` we allow the callback function to return `undefined` to allow filtering out values while they are grouped. Granted, this matters more for `groupBy`, where we can skip building a large array just to throw it away, a problem that doesn't manifest itself here, where we only keep a `number` per key, but it's still useful to keep the size of the object small if a lot of keys aren't needed.

To make the API surface more consistent let's support this here too.

```suggestion
  fn: (value: T, index: number, data: ReadonlyArray<T>) => K | undefined,
```